Code/Framework/AzToolsFramework/Tests/PaintBrush/PaintBrushBaseTests.cpp

```diff
@@ -0,0 +1,249 @@
+- Begin/End Paint Mode
+- Begin/End Brush Stroke
+- Use Eyedropper
+*/
+
+namespace UnitTest
+{
+    class PaintBrushTestFixture : public ScopedAllocatorSetupFixture
+    {
+    public:
+
+        // Some common default values that we'll use in our tests.
+        const AZ::EntityComponentIdPair EntityComponentIdPair{ AZ::EntityId(123), AZ::ComponentId(456) };
+        const AZ::Vector3 TestBrushCenter{ 10.0f, 20.0f, 30.0f };
+        const AZ::Vector3 TestBrushCenter2d{ 10.0f, 20.0f, 0.0f };  // This should be the same as TestBrushCenter, but with Z=0
+        const AZ::Color TestColor{ 0.25f, 0.50f, 0.75f, 1.0f };
+
+        // Useful helpers for our tests
+        AzToolsFramework::PaintBrushSettings m_settings;
+    };
+
+    TEST_F(PaintBrushTestFixture, TrivialConstruction)
+    {
+        // We can construct / destroy a PaintBrush without any errors.
+
+        AzToolsFramework::PaintBrush paintBrush(EntityComponentIdPair);
+    }
+
+    TEST_F(PaintBrushTestFixture, BeginAndEndPaintModeMustBePairedCorrectly)
+    {
+        // BeginPaintMode() / EndPaintMode() can only be called in the correct order in pairs.
+        // They will error if they're called in the wrong order or more than once in a row.
+
+        AzToolsFramework::PaintBrush paintBrush(EntityComponentIdPair);
+
+        // End before Begin should assert.
+        AZ_TEST_START_TRACE_SUPPRESSION;
+        paintBrush.EndPaintMode();
+        AZ_TEST_STOP_TRACE_SUPPRESSION(1);
+
+        paintBrush.BeginPaintMode();
+
+        // Calling Begin twice should assert.
+        AZ_TEST_START_TRACE_SUPPRESSION;
+        paintBrush.BeginPaintMode();
+        AZ_TEST_STOP_TRACE_SUPPRESSION(1);
+
+        paintBrush.EndPaintMode();
+
+        // Calling End twice should assert.
+        AZ_TEST_START_TRACE_SUPPRESSION;
+        paintBrush.EndPaintMode();
+        AZ_TEST_STOP_TRACE_SUPPRESSION(1);
+    }
+
+    TEST_F(PaintBrushTestFixture, IsInPaintModeFunctionsCorrectly)
+    {
+        // IsInPaintMode() correctly identifies when we're between a BeginPaintMode() and EndPaintMode() call.
+
+        AzToolsFramework::PaintBrush paintBrush(EntityComponentIdPair);
+
+        EXPECT_FALSE(paintBrush.IsInPaintMode());
+        paintBrush.BeginPaintMode();
+        EXPECT_TRUE(paintBrush.IsInPaintMode());
+        paintBrush.EndPaintMode();
+        EXPECT_FALSE(paintBrush.IsInPaintMode());
+    }
+
+    TEST_F(PaintBrushTestFixture, BeginAndEndBrushStrokeMustBePairedCorrectlyInsidePaintMode)
+    {
+        // BeginBrushStroke() / EndBrushStroke() can only be called in the correct order in pairs.
+        // They will error if they're called in the wrong order, more than once in a row, or outside of BeginPaintMode().
+
+        AzToolsFramework::PaintBrush paintBrush(EntityComponentIdPair);
+
+        // BeginBrushStroke without BeginPaintMode should assert.
+        AZ_TEST_START_TRACE_SUPPRESSION;
+        paintBrush.BeginBrushStroke(m_settings);
```

Comment:
suggestion: Enforcing this strict API usage does make a lot of sense but I wanted to mention one term I came across while looking at another library ([flecs](https://flecs.docsforge.com/master/manual/#idempotence)) called '[idempotence](https://en.wikipedia.org/wiki/Idempotence)'. It means invoking an operation once has the same effect as invoking an operation multiple times, which would mean if you call `begin` a second time before calling `end` it would basically be a noop and not be an error. Maybe this is something you definitely don't want but it just made me think of it and I wanted to share it just as food for thought as a slightly different philosophy (likely with its own downsides/drawbacks as well).

Reply:
Oooh, this is one I feel really strongly about, so I'll skip the suggestion. One of my core design principles is "symmetry". I won't create a Begin() without an End(), a Startup() without a Shutdown(), etc. In the same vein, I do my best to ensure that people have paired, symmetric calls to my APIs. I've generally found it to be an anti-pattern to allow asymmetric calls, and it I believe it to be an overall disservice to the API user to allow this. Most of the time, from what I've seen, asymmetric calls are a result of incorrect API usage, they cause less predictable resource usage and lifetimes, and they're really hard to fix at the point that they start really causing problems.

There are definitely some exception cases every now and again where you need to allow for _nested_ Begin/End calls, but even in those cases I've still lightly enforced pairing with a counter. Though more often than not even in these cases you actually _can_ redesign to avoid nested calls by using keys or identifiers.

Reply:
Totally valid points and I tend to agree with you, just seeing the way this is structured made me think of the idempotence concept and I wanted to mention it just as a philosophical alternative. Please do leave as is, it's just interesting these different viewpoints exist out in the wild 🙂